packages/javascript-decorators/src/memoize.ts

```diff
@@ -0,0 +1,73 @@
+interface MemoizeMap<T, U> {
+  get(key: T): U;
+  has(key: T): boolean;
+  set(key: T, value: U): MemoizeMap<T, U>;
+}
+
+type Resolver<T extends Function> = T extends (...args: infer T) => any
+  ? T
+  : any;
+
+export const MAX_MAP_ENTRIES = 50;
+
+export default function memoize<F extends Function>(
+  resolver?: Resolver<F>,
+): MethodDecorator {
+  return function<F>(
+    _target: Object,
+    _propertyKey: string | symbol,
+    descriptor: TypedPropertyDescriptor<F>,
+  ) {
+    const method: F | undefined = descriptor.value;
+
+    if (!method || !(method instanceof Function)) {
+      return descriptor;
+    }
+
+    descriptor.value = memoized(method, resolver);
+    return descriptor;
+  };
+}
+
+function memoized<F extends Function>(method: F, resolver?: Resolver<F>): F {
```

Comment:
I would export this as sell (maybe as `memoizeFunction`), since it is useful on its own for functions that aren't methods